angular/projects/spark-angular/src/lib/components/sprk-autocomplete/sprk-autocomplete.component.ts

```diff
@@ -0,0 +1,398 @@
+  ElementRef,
+  AfterViewInit,
+  QueryList,
+} from '@angular/core';
+import { SprkInputDirective } from '../../directives/inputs/sprk-input/sprk-input.directive';
+import { SprkAutocompleteResultsDirective } from './sprk-autocomplete-results/sprk-autocomplete-results.directive';
+import { SprkAutocompleteResultDirective } from './sprk-autocomplete-result/sprk-autocomplete-result.directive';
+import { uniqueId } from 'lodash';
+
+@Component({
+  selector: 'sprk-autocomplete',
+  template: `<ng-content></ng-content>`,
+})
+export class SprkAutocompleteComponent
+  implements AfterContentInit, AfterViewInit {
+  constructor(private ref: ElementRef, private renderer: Renderer2) {}
+  /**
+   * This component expects a child element
+   * with the `sprkInput` directive.
+   */
+  @ContentChild(SprkInputDirective, { static: false })
+  input: SprkInputDirective;
+
+  /**
+   * This component expects a child element
+   * with the `sprkAutocompleteResults` directive.
+   */
+  @ContentChild(SprkAutocompleteResultsDirective, {
+    static: false,
+    read: ElementRef,
+  })
+  results: ElementRef;
+
+  /**
+   * This component expects the `sprkAutocompleteResults` child
+   * to contain elements with the `sprkAutocompleteResult` directive.
+   */
+  @ContentChildren(SprkAutocompleteResultDirective, { descendants: true })
+  resultItems: QueryList<SprkAutocompleteResultDirective>;
+
+  /**
+   * If `true`, the Autocomplete results will be visible when the component loads.
+   */
+  @Input()
+  isOpen = false;
+
+  /**
+   * This event will be emitted after the Autocomplete results are opened.
+   */
+  @Output()
+  openedEvent = new EventEmitter<any>();
+  /**
+   * This event will be emitted after the Autocomplete results are closed.
+   */
+  @Output()
+  closedEvent = new EventEmitter<any>();
+  /**
+   * This event will be emitted when an Autocomplete result is selected with
+   * a click event or with the Enter key. The event contains the id of the
+   * selected item.
+   */
+  @Output()
+  itemSelectedEvent = new EventEmitter<any>();
+
+  /**
+   * @ignore
+   */
+  @HostListener('document:keydown', ['$event'])
+  onKeydown($event) {
+    if (this.isEscapePressed($event) && this.isOpen) {
+      this.hideResults();
+    }
+
+    if (this.isUpPressed($event) && this.isOpen) {
+      this.retreatHighlightedItem();
+    }
+
+    if (this.isDownPressed($event) && this.isOpen) {
+      this.advanceHighlightedItem();
+    }
+
+    // Only select the list item if the list is open
+    // and an item is highlighted
+    if (
+      this.isEnterPressed($event) &&
+      this.isOpen &&
+      this.highlightedIndex !== -1
+    ) {
+      this.selectHighlightedListItem();
+    }
+  }
+
+  /**
+   * @ignore
+   */
+  @HostListener('document:click', ['$event'])
+  onDocumentClick(event): void {
+    if (
+      this.ref.nativeElement &&
+      !this.ref.nativeElement.contains(event.target) &&
+      this.isOpen
+    ) {
+      this.hideResults();
```

Comment:
what happens when the if statement returns false? Nothing? do we need a `return`?

Reply:
Nothing should happen when the if returns false. I don't think we need a `return`. The function returns on its own when it gets to the end of the function and we don't need to return any specific value to the caller.